### PR TITLE
Add Razer BlackWidow Elite support

### PR DESCRIPTION
**Pull request by beranat**
[Razer BlackWidow Elite keyboard](https://www.razer.com/gaming-keyboards-keypads/razer-blackwidow-elite) support in driver and daemon.

Change log:

- Fully supported effects None, Static, Breath, Spectrum, Reactive and Starlight
- Partially supported effect Wave - 'Left-to-Right' only
- All multimedia button and spinner also works (generic (device_type=0:0) and driver (device_type=3:0) modes)
- Added scripts/driver/blackwidow_elite/test.py for checking
- Added writing possibility to 'test' device - it sends command direct to keyboard

### CODE REVIEW

**Comment by beranat** [comment]
*NOTE*: Razer Synapse 3 generates the effects manually (it sends 'set color for every key') so some effects or its parts may be not implemented by hardware and not works at some firmwares.

**Comment by z3ntu** [comment]
What's the status? Do all the effects work?

**Comment by beranat** [comment]
Hello and sorry for long response - I test hard and face that the following features do not work:
* Wave Effect Right-to-Left (but Left-to-Right works) 
* Multimedia button play/pause (in Driver Mode)
* Volume control (in Driver Mode)

**Comment by z3ntu** [comment]
Try writing directly to the `matrix_effect_wave` file in `/sys/bus/hid/drivers/razerkbd/0*/matrix_effect_wave` (I think `echo -ne '1' > matrix_effect_wave` should work for that) and try the values 0 and 1. Does that change the direction of the effect?

**Comment by beranat** [comment]
Dear **z3ntu**, as I wrote earlier - Wave from Left to Right works well (it starts by sending '1'), but sending '0' for right-to-left - does not start effect.

**Comment by MLDMoritz** [comment]
Finally, if you need more tests I'm willing to help.

**Comment by beranat** [comment]
Dear **MLDMoritz**, I would like to thank you for help offer, but I have faced with volume control issue  and I will gladly accept your help in testing after after fixing the problem (it needs a evening).

**Comment by beranat** [comment]
Dear **@MLDMoritz** and **@z3ntu** , I have fixed volume control and would like to thank you for help and assist and rev#f523b98 it looks like release candidate. 

Change log: 
* Fully supported effects None, Static, Breath, Spectrum, Reactive and Starlight 
* Partially supported effect Wave - 'Left-to-Right' only
* All multimedia button and spinner also works (generic (device_type=0:0) and driver (device_type=3:0) modes)
* Added scripts/driver/blackwidow_elite/test.py for checking
* Added writing possibility to 'test' device - it sends command direct to keyboard

**Comment by MLDMoritz** [comment]
I can't seem to find any more issues, works fine for me. Could you check if it connects with polychromatic.app ?

**Comment by beranat** [comment]
Dear **MLDMoritz**, Polychromatic Controller works except one - wrong keyboard layout:
* Razer Logo uses Right-Alt
* Right-Alt n/a
* Multimedia Buttons 
* Space-key 
But these issues should be fixed in Polychromatic code base, because custom colour test can activate all enumerated buttons.

**Comment by ntedgi** [comment]
hi after installing 
@beranat drivers this is the dmesg output on linux mint 

Screenshot from 2019-04-07 12-17-13

tested and works on Linux mint 19 with polychromatic thanks a lot!!
Screenshot from 2019-04-07 12-43-43

let me know if you want to make another test or if i can help

**Comment by beranat** [comment]
Dear **ntedgi**, I would like to thank you for feedback. 
The DMESG's mesage is absolutely correct and notifies about USB-device with 3 HID-devices (Elite keyboard really has been implemented as 3 ones).

**Comment by beranat** [comment]
[BlackWidow Elite support patch for polychromatic.app](https://github.com/polychromatic/polychromatic/pull/200) has been asked for adding to theirs repo as I have promised before.

screen

**Comment by MLDMoritz** [comment]
Awesome! @z3ntu, @terrycain  anything else needed for a merge?

**Comment by mkhattat** [comment]
@beranat Thanks it works great (except wave effect right-to-left). 
Just one question: This keyboard has an on-board memory so basically you can save a profile. Dose `matrix_custom_frame` write to the memory or is it just some sort of register? I ask it because I wanted to make custom light show using `matrix_custom_frame` but I don't want to wear the on-board memory out.

File: driver/razerkbd_driver.h [hunk 1]
```diff
@@ -88,7 +89,7 @@
 #define RAZER_STEALTH_ROW_LEN 0x10
 #define RAZER_STEALTH_ROWS_NUM 6
 
-
+#define RAZER_BLACKWIDOW_ELITE_TRANSACTION_ID 0x1F
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: It'd be better if you'd just use 0x1F instead of the macro

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -2414,11 +2556,16 @@ MODULE_DEVICE_TABLE(hid, razer_devices);
...
     .remove = razer_kbd_disconnect,
 
     .event = razer_event,
-    .raw_event = razer_raw_event,
+    .raw_event = razer_raw_event
 };
 
 module_hid_driver(razer_kbd_driver);
+
+// vim: tabstop=4 shiftwidth=4 softtabstop=4 expandtab colorcolumn=96 :
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Remove the modeline please

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -1748,6 +1870,23 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
...
+ * Set static hid-events translation map
+ *
+ * [BWE-Wheel] 'BlackWidow Elite' generates wheel-events for volume control spinner
+ */
+static int razer_kbd_input_mapping(struct hid_device *hdev, struct hid_input *hi, struct hid_field *field,
+                                   struct hid_usage *usage, unsigned long **bit, int *max)
+{
+
+    // [BWE-Wheel]
+    if (USB_DEVICE_ID_RAZER_BLACKWIDOW_ELITE == hdev->product && HID_TYPE_USBMOUSE == hdev->type && HID_GD_WHEEL == usage->hid) {
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Please rather do `if (variable == constant)` instead of the way it's currently

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -1298,7 +1390,7 @@ static ssize_t razer_attr_write_set_brightness(struct device *dev, struct device
 }
 
 /**
- * Read device file "macro_mode"
+ * Read device file "set_brightness"
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: The sysfs file is called `matrix_brightness` (if you're at it please also fix the comment at the write_set_brightness function)

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -2414,11 +2556,16 @@ MODULE_DEVICE_TABLE(hid, razer_devices);
...
     .id_table = razer_devices,
+
+    .input_mapping = razer_kbd_input_mapping,
+
     .probe = razer_kbd_probe,
     .remove = razer_kbd_disconnect,
 
     .event = razer_event,
-    .raw_event = razer_raw_event,
+    .raw_event = razer_raw_event
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: leave the comma here

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (z3ntu) [reply driver/razerkbd_driver.c]: Also can you check here for the Razer Huntsman Elite? See #803

Comment (beranat) [reply driver/razerkbd_driver.c]: Dear **@z3ntu**, Yes, I can check this issue and think the problem is the same and after answering [question about multimedia buttons](https://github.com/openrazer/openrazer/issues/803#issuecomment-483223819) I try to implement that keyboard full support too.

Comment (beranat) [reply driver/razerkbd_driver.c]: Dear **@z3ntu**, (variable == constant) is not a good practice - the code must be as issue persistent as possible and misprint *constant = variable* vs *variable == constant* will be detected at first compilation process, but *Legacy for the Future* - you are project owner so I will try to reverse if-expressions when I have a spare time.

File: driver/razerkbd_driver.h (hunk 1 above)
Comment (beranat) [reply driver/razerkbd_driver.h]: "Magic" value in 13 places could not be better then in one, but I will pay attention to your comment.

File: driver/razerkbd_driver.c (hunk 2 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: Ok, I will remove this modeline

File: driver/razerkbd_driver.c (hunk 4 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: Yes, thank you

File: driver/razerkbd_driver.c (hunk 5 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: Dear **@z3ntu**, I have not this comma at c4a29df (blackwiNdow fix), may be in some transitional commit

File: driver/razerkbd_driver.h (hunk 1 above)
Comment (z3ntu) [reply driver/razerkbd_driver.h]: It's not really a magic value, it has some structure behind it but it isn't important to know. If you really want to use a constant, rename it to `RAZER_TRANSACTION_ID_1F` or something like that as it's probably also used by other devices

Comment (beranat) [reply driver/razerkbd_driver.h]: Ok, on the way

**Comment by TorkStro** [comment]
hi I wana test this how do i install it?

**Comment by Severus157** [comment]
Great to see the Support is coming along for supporting the Keyboard, I just got my Keyboard today and am looking forward to be able to change the lighting effects on my PC.
Is there a new Version of Openrazer coming soon where the Support is included? I want to avoid compiling it myself if possible :)

**Comment by fiercebrake** [comment]
> hi I wana test this how do i install it?
@TorkStro you could follow the instructions at "https://github.com/openrazer/openrazer/wiki/Building-a-package" hopefully, your distro will be listed there.

**Comment by beranat** [comment]
Sorry, some crash on ssd! Will update soon

**Comment by beranat** [comment]
Codebase restored, there are new commits, but 'Reopen PR' button is still not availbe.

**Comment by mestaburrito** [comment]
What's the update on this? Can we drop the instructions on how to at least get some of it compatible with the polychromatic controller software?

**Comment by micach** [comment]
please check my comment in https://github.com/openrazer/openrazer/issues/803#issuecomment-495198136

**Comment by DangerBlack** [comment]
# Installation

How to use Razer BlackWidow Elite on Linux mint.

# Install openrazer

```bash
sudo apt-get install build-essential
sudo apt-get install dh-python
git clone https://github.com/beranat/openrazer
cd openrazer
sudo apt install debhelper linux-headers-generic python3-setuptools lsb-release dkms # build dependencies
./scripts/build_debs.sh
sudo dpkg -i ./dist/*.deb
```
# Install polychromatic
```bash
sudo add-apt-repository ppa:polychromatic/stable
sudo apt update
sudo apt install polychromatic
```

# Fix polychromatic

```bash
git clone https://github.com/beranat/polychromatic.git
cd polychromatic
sudo cp data/mapping/blackwidow-elite-keyboard-layout.svg /usr/share/polychromatic/mapping/
sudo cp data/pages/profile_editor.js /usr/share/polychromatic/pages/
sudo cp polychromatic-controller /usr/bin/
cd ../
rm -r polychromatic/
```


In order to make everything work you need to reboot!!!

## Reference

* https://github.com/openrazer/openrazer/wiki/Building-a-package

**Comment by z3ntu** [comment]
Please move the discussion to #831

**Pull request closed without merging**